README.md
# 小工具

## 背景

操作类工作均有规律可寻，员工按照“操作手册”即可完成工作。其中一些工作内容逻辑清晰，规则健全，若配合 IT 技术，可实现自动化。

本仓库为方便工作开发的各种小工具，将重复性工作、人工操作效率低、易出错的工作，用 IT 技术辅助来完成。每个工具均可实现某项工作全自动或半自动作业，实现 IT 换人。这将大大提高效率，减少员工机械重复的工作，减少人工失误，让员工将时间精力投入到其他工作中，实现降本增效。

## 目标

- 傻瓜式操作，一键运行或输入变量即可得到结果
- 适配多种场景，易扩展
- 效率高、准确性高、用户体验好

## 一致性检查工具

简单概括需求，就是 1 个原始数据，2 个需要对比的数据，3 个表格均为 excel 文件，表的格式、包含的字段、字段的标题都不相同，但是字段标题存在某种对应关系或计算关系，现需要对多个字段进行一致性对比。理想情况想要的结果是能知道哪些行不一致，具体不一致的内容是哪一列，特点是数据量大，需要经常对比。[使用说明](https://github.com/thianda/xda-tools/tree/master/ip-contrast) 

**实现思路**

1. 整理原始数据：剔除无用数据列、筛选掉无效的行，准确识别字段的标题
2. 对比并标记不一致。
3. 生成友好的适合人阅读的结果

## IPv6 地址拆分工具

目前处于设计阶段，具体实现的需求详见 [这里](https://github.com/thianda/xda-tools/tree/master/ipv6-split/)。



ip-contrast/CHANGELOG.md
# 一致性检查工具更新历史

### v0.8.7 (2019-03-07)

- 修复丢失多行标题情况下可能丢失第一行数据的 bug

### v0.8.6 (2019-02-28)

- 修复读取多个 sheet 页时会崩溃的 bug。`（影响版本 since v0.8.3）`
- 修正了一致性对比公式。
- `所属地市`过滤字段非必须。
- 智能识别表格标题中`*`的干扰。
- 修复 linux 下的输出 bug。

**TODO**

- 多文件多 sheet 情况下尝试多线程同时解析
- 修改 excel 公式以匹配成段的地址


### v0.8.3 (2019-02-26)

- 添加`所属地市`过滤字段。
- 自动过滤部分非有效数据。
- 优化处理逻辑，进一步防止异常闪退。

### v0.8.2 (2019-02-24)

- 优化运行进度显示
- 更新 `README.md`
- 适配 linux，支持在 linux 服务器上后台运行。

### v0.8.1 (2019-02-23)

- 取消内存优化，用于提升运行效率
> 注释掉 `del()`

### v0.8.0 (2019-02-23)

- 自动过滤 ipv6。
- 同类型文件识别多个文件多个sheet。
- 运行时实时显示进度。
- 拒绝崩溃，友好显示错误信息。
- 自动识别 before 参数。
- 预处理时间字段为：`%Y-%m-%d`，移除 `%H:%M:%S`。
- 内存优化，降低运行时内存占用。该功能使运行效率降低，总运行时间扩大20倍。


### v0.7.1 (2019-02-22)

- 优化处理逻辑，加速运行效率。
- 修复已知bug，避免崩溃。

### v0.7.0 (2018-12-23)

- 自动判断检查内容。
- 优化命令行输出。
- 优化输出的结果文件。
- 系统导出数据异常时，友好提示。

### v0.6.0 (2018-12-12)

初步可以输出结果文件。并自动打开到输出目录。

### v0.5.0 (2018-12-10)

可以将对比文件基于生成中间文件。

优化结果的显示格式。

ip-contrast/ip-contrast.v0.9.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from datetime import datetime
import configparser
import os
import platform
import sys
import ctypes
import numpy
import pandas
import traceback


class ip_contrast(object):
    """ IP 信息对比"""

    def __init__(self):
        self.t0 = datetime.now()
        self.__version__ = '0.9.0'
        self.config = configparser.ConfigParser()
        self.configFileName = 'config_{}.ini'.format(self.__version__)
        self.DEBUG_FILE = 'debug_log.txt'

        print('****欢迎使用一致性检查工具 {}\n'.format(self.__version__))
        print(self.now(), '**开始运行')
        """初始化配置"""
        if os.path.exists(self.configFileName):
            if not self.checkConfig():
                print('Warning:', '配置有误，已恢复默认', self.configFileName)
                self.writeConfig()
        else:
            print('Info:', '配置已更新', self.configFileName)
            # 释放默认配置
            self.writeConfig()
        print('当前加载的配置：', self.configFileName)
        self._matchFileNames()

    def now(self):
        """当前时间的字符串"""
        return datetime.strftime(datetime.now(), '%Y-%m-%d %H:%M:%S')

    def checkConfig(self):
        """检查配置格式是否正确"""
        self.config.read(self.configFileName)
        sections = self.config.sections()
        sectionLens = len(sections)
        if sectionLens < 5:
            return False
        # 每个 section 中的 options 的个数必须是相同的
        SETTING_LENS = 2
        for i in range(SETTING_LENS, sectionLens):
            if i < sectionLens - SETTING_LENS:
                n1 = len(self.config.options(sections[i]))
                n2 = len(self.config.options(sections[i + 1]))
                if n1 == 0 | n1 != n2:
                    return False
        # 可进一步判断 options 名是否一致
        pass
        return True

    def writeConfig(self):
        config = self.config
        config['common'] = {
            'version': self.__version__
        }
        config['对比文件名'] = {
            '省内资管': 'IP地址.',
            '集团': '-IP地址-',
            '工信部备案': 'fpxxList',
        }
        config['省内资管'] = {
            'tag1': '所属地市',
            'ip': 'IP地址',
            'starttime': '分配使用时间',
            'field3': '联系人姓名(客户侧)',
            'field4': '联系电话(客户侧)',
            'field5': '单位详细地址',
            'field6': '联系人邮箱(客户侧)',
            'field7': '单位名称/具体业务信息',
        }
        config['集团'] = {
            'tag1': '所属地市',
            'ip': '网段名称',
            'starttime': '分配使用时间',
            'field3': '联系人姓名(客户侧)',
            'field4': '联系人电话(客户侧)',
            'field5': '单位详细地址',
            'field6': '联系人邮箱(客户侧)',
            'field7': '单位名称/具体业务信息',
        }
        config['工信部备案'] = {
            'tag1': '所属地',
            'ip': '起始IP;终止IP',
            'starttime': '分配日期',
            'field3': '联系人姓名',
            'field4': '联系人电话',
            'field5': '单位详细地址',
            'field6': '联系人电子邮件',
            'field7': '使用单位名称',
        }
        with open(self.configFileName, 'w') as configFile:

            title = '''# Author: Xianda

    # 本配置为一致性检查工具的配置
    # 如删除本配置文件，重新生成的配置文件即为默认配置
    # 如需修改配置，修改本文件后直接保存即可

    ######

    # before 表示数据的起始行（列名占用的行数）。现已自动识别（配置无效）
    # 若 IP 地址字段分为起始IP、终止IP的，`ip` 字段中用“;”(英文分号)隔开
    # 程序会依次对 field 字段进行对比并输出对比结果
    # field 字段有变化可直接在此增删改，满足一一对应即可



    '''
            configFile.write(title)
            config.write(configFile)

    def _matchFileNames(self):
        """匹配导出数据文件名"""
        fileName = {}
        self.config.read(self.configFileName)
        for option in self.config.options('对比文件名'):
            value = self.config.get('对比文件名', option)
            fileName[option] = []
            for f in os.listdir():
                if value in f and not '~' in f:
                    fileName[option].append(f)
            if not fileName[option]:
                fileName.pop(option)
                print('Warning:', '未找到', option, '的导出数据，将不进行该数据的一致性检查。')
        self.fileName = fileName

    def recognizeOptions(self, k, xls):
        """识别配置"""
        fields = self.config.options(k)
        options = {}
        options['fieldCols'] = {}
        options['ipCols'] = []
        colNames = []
        ipNames = []
        # 默认获取第一个 sheet 的第一行做为列名所在的行
        Row0 = xls.sheet_by_index(0).row_values(0)
        # 遍历配置文件中要对比的列名 记录要对比的字段所在的列
        for field in fields:
            configValue = self.config.get(k, field)
            colFounded = False
            # 遍历导出数据的第一行单元格
            for i in range(len(Row0)):
                cellValue = str(Row0[i].strip().strip('*'))
                if cellValue in configValue:  # 为了满足： `起始IP` in `起始IP;终止IP`
                    colFounded = True
                    if 'ip' == field:
                        # 记录 ip 所在的列
                        options['ipCols'].append(i)
                        ipNames.append(Row0[i])
                    else:
                        # 记录要对比的字段所在的列
                        options['fieldCols'][field] = i
                        colNames.append(Row0[i])
                    # 此处不能加 break，否则匹配到`起始IP`即跳出 for 循环，无法匹配`结束IP`
            # 若根据配置未找到对应的列：给出提示，结束运行
            if not colFounded:
                if not field in ['tag1']:
                    print('在{}中，未找到数据列：【{}】，工具即将退出。'.format((k, configValue))
                    pause()
                    exit()
        self.options, self.colNames, self.ipNames=options, colNames, ipNames
        return options, colNames, ipNames

    def run(self):
        """执行对比"""
        # self.recognizeOptions()


if 'Windows' in platform.system():
    __stdInputHandle=-10
    __stdOutputHandle=-11
    __stdErrorHandle=-12
    __foreGroundBLUE=0x09
    __foreGroundGREEN=0x0a
    __foreGroundRED=0x0c
    __foreGroundYELLOW=0x0e
    stdOutHandle=ctypes.windll.kernel32.GetStdHandle(__stdOutputHandle)

    def setCmdColor(color, handle=stdOutHandle):
        return ctypes.windll.kernel32.SetConsoleTextAttribute(handle, color)

    def resetCmdColor():
        setCmdColor(__foreGroundRED | __foreGroundGREEN | __foreGroundBLUE)

    def printBlue(msg):
        setCmdColor(__foreGroundBLUE)
        sys.stdout.write(msg + '\n')
        resetCmdColor()

    def printGreen(msg):
        setCmdColor(__foreGroundGREEN)
        sys.stdout.write(msg + '\n')
        resetCmdColor()

    def printRed(msg):
        setCmdColor(__foreGroundRED)
        # sys.stdout.write(msg + '\n')
        print(msg)
        resetCmdColor()

    def printYellow(msg):
        setCmdColor(__foreGroundYELLOW)
        sys.stdout.write(msg + '\n')
        resetCmdColor()

    def pause():
        os.system('pause')

    def locateFile(file):
        print('\n- 即将跳转到输出结果 -\n')
        pause()
        os.system('explorer /select, ' + file)

else:
    STYLE={
        'fore': {
            'red': 31,
            'green': 32,
            'yellow': 33,
            'blue': 34,
        }
    }

    def UseStyle(msg, mode='', fore='', back='40'):
        fore='%s' % STYLE['fore'][fore] if fore in STYLE['fore'] else ''
        style=';'.join([s for s in [mode, fore, back] if s])
        style='\033[%sm' % style if style else ''
        end='\033[%sm' % 0 if style else ''
        return '%s%s%s' % (style, msg, end)

    def printRed(msg):
        print(UseStyle(msg, fore='red'))

    def printGreen(msg):
        print(UseStyle(msg, fore='green'))

    def printYellow(msg):
        print(UseStyle(msg, fore='yellow'))

    def printBlue(msg):
        print(UseStyle(msg, fore='blue'))

    def pause():
        pass

    def locateFile(file):
        print('\n- 结果文件保存为：{}'.format(file))
        print('刷新页面可到 `__output` 文件夹中下载查看。')

if __name__ == '__main__':
    ip_tables=ip_contrast()
    try:
        ip_tables.run()
    except Exception as err:
        printRed('Error:')
        traceback.print_exc(file=open(ip_tables.DEBUG_FILE, 'w'))
        printRed(err)
        printBlue('出错啦。请反馈目录中的 debug_log.txt 文件内容')

    t=datetime.now() - ip_tables.t0
    print('\n执行用时：{:.4f} s'.format(t.total_seconds()))
    locateFile(ip_tables.result)


ip-contrast/ip-contrast.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import platform
import configparser
import os
import xlrd
from datetime import datetime
import openpyxl
import re
import traceback
# import shutil

__version__ = '0.8.7'
configFileName = 'config_%s.ini' % __version__
DEBUG_FILE = 'debug_log.txt'
config = configparser.ConfigParser()
t0 = datetime.now()


# 当前时间的字符串
def now():
    return datetime.strftime(datetime.now(), '%Y-%m-%d %H:%M:%S')


# ip 转换： int 转 str
def int2ip(x):
    return '.'.join([str(int(x / (256 ** i) % 256)) for i in range(3, -1, -1)])


# ip 转换： str 转 int
def ip2int(x):
    try:
        return sum([256 ** j * int(i) for j, i in enumerate(x.split('.')[::-1])])
    except:
        return False


# ip 字符串 10.10.10.10/32 转成数组: ip/mask/ip_start/ip_end
def ipParse(ipStr):
    if ipStr == '':
        return [None, None, None, None]
    maskLen = 32
    if '/' in ipStr:
        ipStr, maskLen = ipStr.split('/')
    # print(ipStr, maskLen)
    ipInt = ip2int(ipStr)
    if maskLen == 32:
        return [ipInt, 4294967295, ipInt, ipInt]
    mask = 0xFFFFFFFF << (32 - int(maskLen))
    # print('mask:', mask)
    ipStart = ipInt & mask & 0xFFFFFFFF
    ipEnd = ipInt | ~mask & 0xFFFFFFFF
    return [ipInt, mask, ipStart, ipEnd]


# ip, mask 转成 字符串 10.10.10.10/32
def ipImport(ipInt, mask):
    pass


# 检查配置格式是否正确
def checkConfig(configFileName):
    global config
    config.read(configFileName)
    # _version = config.get('common', 'version', fallback=None)
    # if not _version:
    #     return False
    # _version = [int(i) for i in _version.split('.')]
    # _current_ver = [int(i) for i in __version__.split('.')]
    # if _version < _current_ver:
    #     return False
    sections = config.sections()
    sectionLens = len(sections)
    if sectionLens < 5:
        return False
    # 每个 section 中的 options 的个数必须是相同的
    SETTING_LENS = 2
    for i in range(SETTING_LENS, sectionLens):
        if i < sectionLens - SETTING_LENS:
            n1 = len(config.options(sections[i]))
            n2 = len(config.options(sections[i + 1]))
            if n1 == 0 | n1 != n2:
                return False
    # 可进一步判断 options 名是否一致
    pass
    return True


# 释放默认配置
def writeConfig(configFileName):
    global config
    config['common'] = {
        'version': __version__
    }
    config['对比文件名'] = {
        '省内资管': 'IP地址.',
        '集团': '-IP地址-',
        '工信部备案': 'fpxxList',
    }
    config['省内资管'] = {
        'tag1': '所属地市',
        'ip': 'IP地址',
        'starttime': '分配使用时间',
        'field3': '联系人姓名(客户侧)',
        'field4': '联系电话(客户侧)',
        'field5': '单位详细地址',
        'field6': '联系人邮箱(客户侧)',
        'field7': '单位名称/具体业务信息',
    }
    config['集团'] = {
        'tag1': '所属地市',
        'ip': '网段名称',
        'starttime': '分配使用时间',
        'field3': '联系人姓名(客户侧)',
        'field4': '联系人电话(客户侧)',
        'field5': '单位详细地址',
        'field6': '联系人邮箱(客户侧)',
        'field7': '单位名称/具体业务信息',
    }
    config['工信部备案'] = {
        'tag1': '所属地',
        'ip': '起始IP;终止IP',
        'starttime': '分配日期',
        'field3': '联系人姓名',
        'field4': '联系人电话',
        'field5': '单位详细地址',
        'field6': '联系人电子邮件',
        'field7': '使用单位名称',
    }
    with open(configFileName, 'w') as configFile:
        configFile.write(
            '''# Author: Xianda

# 本配置为一致性检查工具的配置
# 如删除本配置文件，重新生成的配置文件即为默认配置
# 如需修改配置，修改本文件后直接保存即可

######

# before 表示数据的起始行（列名占用的行数）。现已自动识别（配置无效）
# 若 IP 地址字段分为起始IP、终止IP的，`ip` 字段中用“;”(英文分号)隔开
# 程序会依次对 field 字段进行对比并输出对比结果
# field 字段有变化可直接在此增删改，满足一一对应即可



''')
        config.write(configFile)


# 初始化配置
def initConfig():
    global configFileName
    if os.path.exists(configFileName):
        if not checkConfig(configFileName):
            print('Warning:', '配置有误，已恢复默认', configFileName)
            writeConfig(configFileName)
    else:
        print('Info:', '配置已更新', configFileName)
        # 释放默认配置，开始对比数据
        writeConfig(configFileName)
    print('当前加载的配置：', configFileName)
    return contrast()


# 读取配置
def readConfig(section, option):
    global config
    global configFileName
    config.read(configFileName)
    value = config.get(section, option)
    return value


# 匹配导出数据文件名
def matchedFileName():
    global config
    global configFileName
    fileName = {}
    config.read(configFileName)
    for option in config.options('对比文件名'):
        value = config.get('对比文件名', option)
        fileName[option] = []
        for f in os.listdir():
            if value in f and not '~' in f:
                fileName[option].append(f)
        if not fileName[option]:
            fileName.pop(option)
            print('Warning:', '未找到', option, '的导出数据，将不进行该数据的一致性检查。')
    return fileName


# 识别配置
def recognizeOptions(k, xls):
    global config
    fields = config.options(k)
    options = {}
    options['fieldCols'] = {}
    options['ipCols'] = []
    colNames = []
    ipNames = []
    # 默认获取第一个 sheet 的第一行做为列名所在的行
    Row0 = xls.sheet_by_index(0).row_values(0)
    # 遍历配置文件中要对比的列名 记录要对比的字段所在的列
    for field in fields:
        configValue = config.get(k, field)
        colFounded = False
        # 遍历导出数据的第一行单元格
        for i in range(len(Row0)):
            cellValue = str(Row0[i].strip().strip('*'))
            if cellValue in configValue:  # 为了满足： `起始IP` in `起始IP;终止IP`
                colFounded = True
                if 'ip' == field:
                    # 记录 ip 所在的列
                    options['ipCols'].append(i)
                    ipNames.append(Row0[i])
                else:
                    # 记录要对比的字段所在的列
                    options['fieldCols'][field] = i
                    colNames.append(Row0[i])
                # 此处不能加 break，否则匹配到`起始IP`即跳出 for 循环，无法匹配`结束IP`
        # 若根据配置未找到对应的列：给出提示，结束运行
        if not colFounded:
            if not field in ['tag1']:
                print('在%s中，未找到数据列：【%s】，工具即将退出。' % (k, configValue))
                pause()
                exit()
    return options, colNames, ipNames


# 生成中间文件，转换导出数据为每一行一个 ip
def generateTemp(fileName):
    if fileName == {}:
        print('Error:', '未识别到excel文件。')
        pause()
        exit()
    global config
    # 基于 ip 列 生成中间文件，并补充 ipStart、ipEnd 列
    options = {'total': 0, 'current': 0}
    colNames = {}
    ipNames = {}
    sheet = {}
    # 按导出文件名遍历
    _t = datetime.now()
    print(now(), '识别表格配置...')
    for k, vv in fileName.items():
        # 遍历同一类型的所有文件，比如集团导出文件分多个
        isFirstFile = True
        for v in vv:
            xls = xlrd.open_workbook(v)
            for i in xls.sheet_names():
                options['total'] += xls.sheet_by_name(i).nrows
                if isFirstFile:
                    options[k], colNames[k], ipNames[k] = recognizeOptions(
                        k, xls)
                    isFirstFile = False
            xls = None
        # colNames[k] = ['ipStart', 'ipEnd'] + colNames[k]
        colNames[k].extend(['ipStart', 'ipEnd'])
        colNames[k].extend(ipNames[k])
        # del(xls, Row0, field, v)
    t = (datetime.now() - _t).total_seconds()
    print(now(), '表格配置识别完毕。用时 %2.4f 秒。' % t)
    # print('Info:', 'options', options, '\n')
    # print('Info:', 'colNames', colNames, '\n')
    SEP = os.path.sep
    configPath = os.getcwd() + SEP + '..' + SEP + '__output' + SEP
    # tempdir=>C:\\ProgramData
    # tempdir = os.environ.get('ALLUSERSPROFILE')
    # tempdir=>%USERPROFILE%/Local/Temp, C:/Users/XXXX/AppData/Local/Temp
    tempdir = os.environ.get('TEMP')
    if tempdir and os.path.exists(tempdir):
        configPath = tempdir + SEP + 'Xianda' + SEP + 'ipContrast' + SEP
    if not os.path.exists(configPath):
        os.makedirs(configPath)
    # 生成结果
    note = '对比结果在第二个sheet页最后 ' + str(len(colNames) - 1) + ' 列'
    author = '--Xianda'
    # wb = openpyxl.Workbook(write_only=True)
    # ws0 = wb.create_sheet('说明')
    # from openpyxl.worksheet.write_only import WriteOnlyCell
    # from openpyxl.styles import Font
    # cell = WriteOnlyCell(ws0, value=note)
    # cell.font = Font(size=18, color='FF0000')
    # ws0.append([None])
    # ws0.append([None, cell])
    # ws0.append([None])
    # cell = WriteOnlyCell(ws0, value=author)
    # cell.font = Font(size=11, color='8060ee', bold=True)
    # ws0.append([None, None, cell])
    # cell = WriteOnlyCell(ws0, value=str(options))
    # cell.font=Font(size=11, color='666600')
    # ws0.append([None, cell])
    # cell=WriteOnlyCell(ws0, value=str(colNames))
    # cell.font=Font(size=11, color='666600')
    # ws0.append([None, cell])
    wb = openpyxl.Workbook()
    ws0 = wb.active  # 第一个 sheet
    ws0.title = '说明'
    ws0.column_dimensions['B'].width = 120
    ws0['B2'] = note
    ws0['C3'] = author
    line = 6
    count = len(fileName) + 1
    for i in fileName:
        ws0['A' + str(line)] = i
        ws0['B' + str(line)] = str(options[i])
        ws0['A' + str(line + count)] = i
        ws0['B' + str(line + count)] = str(colNames[i])
        line += 1
    # ws.merge_cells('A2:G3')
    del(ws0)
    currDate = datetime.strftime(datetime.now(), '%Y-%m-%d_%H%M%S')
    wrName = configPath + 'results-' + currDate + '.xlsx'
    isFirstSheet = True
    ipv4_reg = r'(25[0-5]|2[0-4]\d|[0-1]?\d?\d)(\.(25[0-5]|2[0-4]\d|[0-1]?\d?\d)){3}'
    for k, vv in fileName.items():
        print('\n%s 开始筛选数据：%s' % (now(), k))
        # 为每个对比文件创建一个 sheet
        ws = wb.create_sheet(k)
        if isFirstSheet:
            extendTitle = ['与' + x + '一致' for x in fileName.keys()]
            extendTitle.pop(0)
            title = colNames[k] + ['预留1', '预留2', 'concatenate'] + extendTitle
        else:
            title = colNames[k]
        ws.append(title)
        ipCols = options[k]['ipCols']
        fieldCols = options[k]['fieldCols']
        # 遍历每个文件
        for v in vv:
            _t = datetime.now()
            xls = xlrd.open_workbook(v, on_demand=True)
            t = (datetime.now() - _t).total_seconds()
            print(now(), '加载文件 %s 用时 %2.4f 秒。解析中...' % (v, t))
            # 获取所有 sheet_names
            sheet[k] = xls.sheet_names()
            # 遍历每个 sheet
            sheets_len = len(sheet[k])
            _nrows = 0
            for i_sheet in range(sheets_len):
                _sheet_name = sheet[k][i_sheet]
                if 'Sheet' in _sheet_name:
                    # 疑似修改导出文件手动创建的 sheet，跳过
                    print('Error: sheet 名称包含 `Sheet`，自动跳过。')
                    continue
                _sheet = xls.sheet_by_name(_sheet_name)
                # 自动识别当前 sheet 的 before
                _invalid = True  # 未识别到有效数据为无效
                for i in range(1, 10):
                    col_ip = _sheet.row_values(i)[ipCols[0]]
                    # if i < 2:
                    #     print(_sheet.row_values(i))
                    #     printYellow(col_ip)
                    is_ipv4 = re.search(ipv4_reg, col_ip)
                    if is_ipv4 or ':' in col_ip:
                        _invalid = False
                        _before = i
                        break
                if _invalid:
                    print(now(), 'Error：未识别到有效数据，已跳过：')
                    printRed('%s => [%s]' % (v, _sheet_name))
                    continue
                options['current'] += _before
                # 遍历每一行数据
                _nrows += _sheet.nrows
                for row in range(_before, _sheet.nrows):
                    options['current'] += 1
                    _curr = options['current']
                    _total = options['total']
                    progress = 100*_curr/_total
                    print('\r%s 解析进度：当前 %s/%s sheet %s/%s，总进度：%s/%s，%.3f%%' %
                          (now(), row+1, _nrows, i_sheet+1, sheets_len, _curr, _total, progress), end='')
                    rowValues = _sheet.row_values(row)
                    ip1Str = rowValues[ipCols[0]]
                    if ':' in ip1Str:
                        # ipv6 暂不对比，过滤本行
                        continue
                    tempRow = []
                    if not isFirstSheet:
                        if rowValues[int(fieldCols['field7'])] == '':
                            # 过滤无用数据(field5 为空)
                            continue
                    currentRow = str(ws.max_row + 1)
                    for i in fieldCols:
                        cellValue = rowValues[int(fieldCols[i])]
                        # 转换 starttime 的格式
                        if i == 'starttime' and len(cellValue) > 10:
                            cellValue = cellValue[:10]
                        tempRow.append(cellValue)
                        # del(cellValue)
                    if len(ipCols) == 1:
                        ip2Str = None
                        if '/32' in ip1Str or not '/' in ip1Str:
                            ip_start = ip2int(ip1Str.split('/')[0])
                            ip_end = ip_start
                        else:
                            _ip_parse = ipParse(ip1Str)
                            ip_start = _ip_parse[2]
                            ip_end = _ip_parse[3]
                    elif len(ipCols) == 2:
                        # 导出数据自带起始 IP 结束IP
                        ip2Str = rowValues[ipCols[1]]
                        ip_start = ip2int(ip1Str)
                        ip_end = ip2int(ip2Str)
                        if ip_start > ip_end:
                            ip_start, ip_end = ip_end, ip_start
                            ip1Str, ip2Str = ip2Str, ip1Str
                    else:
                        # ip 字段个数不是 1 也不是 2
                        print('Warning:', 'IP 列识别错误 %s %s 行' % (k, currentRow))
                        return DEBUG_FILE
                    has_tag = len(options[k]['fieldCols']) > 6
                    if has_tag:
                        formula_strs = '=CONCATENATE(B{0},"-",C{0},"-",D{0},"-",E{0},"-",F{0},"-",G{0})'.format(
                            currentRow)
                    else:
                        formula_strs = '=CONCATENATE(A{0},"-",B{0},"-",C{0},"-",D{0},"-",E{0},"-",F{0}'.format(
                            currentRow)
                    tempRow.extend(
                        [ip_start, ip_end, ip1Str, ip2Str, None, formula_strs])
                    if isFirstSheet:
                        formula = [x for x in fileName.keys()]
                        formula.pop(0)
                        formula_strs = []
                        for x in formula:
                            position = [currentRow, x]
                            has_tag2 = len(options[x]['fieldCols']) > 6
                            if has_tag:
                                position.extend(['H', 'M'])
                            else:
                                position.extend(['G', 'L'])
                            if has_tag2:
                                position.append('H:M')
                            else:
                                position.append('G:L')
                            formula_strs.append(
                                '=VLOOKUP({2}{0},{1}!{4},6,0)={3}{0}'.format(*position))
                        tempRow.extend(formula_strs)
                    # print('tempRow', tempRow) # debug
                    # 过滤客户地址为空的数据
                    # if tempRow[3] == '':
                    #     continue
                    # else:
                    ws.append(tempRow)
                    # del(tempRow, ip_start, ip_end, ip1Str, ip2Str, strings)
                    # gc.collect()
            ws.auto_filter.ref = 'A1:O' + str(ws.max_row)
            _t = datetime.now()
            wb.save(wrName)  # 每读取完一个文件保存一次
            t = (datetime.now() - _t).total_seconds()
            print('\n%s 解析完毕 %s ' % (now(), v))
            print('当前进度已保存到： %s' % wrName)
            print('保存操作用时 %2.4f 秒' % t)
        ws.column_dimensions['A'].width = 15
        ws.column_dimensions['H'].width = 11
        ws.column_dimensions['I'].width = 11
        ws.column_dimensions['J'].width = 20
        ws.column_dimensions['K'].width = 20
        isFirstSheet = False
        # ws.auto_filter.add_sort_condition('G2:G'+str(ws.max_row))
        # del(ws)
    wb.save(wrName)
    return wrName


def contrast():
    # 获取导出文件的文件名
    fileName = matchedFileName()
    # print('Info:','fileName', fileName, '\n')
    tempFile = generateTemp(fileName)
    return tempFile


# unused for now
# def copyfile(srcfile, dstfile):
#     if not os.path.isfile(srcfile):
#         print('%s not exist!' % (srcfile))
#     else:
#         fpath = os.path.split(dstfile)[0]  # 分离文件名和路径
#         if not os.path.exists(fpath):
#             os.makedirs(fpath)  # 创建路径
#         shutil.copyfile(srcfile, dstfile)  # 复制文件


# @测试 configparser
def _test_configparser():
    global config
    config.read(configFileName)
    sections = config.sections()
    for x in sections:
        options = config.options(x)
        for y in options:
            print(config.get(x, y))


if 'Windows' in platform.system():
    import sys
    import ctypes
    __stdInputHandle = -10
    __stdOutputHandle = -11
    __stdErrorHandle = -12
    __foreGroundBLUE = 0x09
    __foreGroundGREEN = 0x0a
    __foreGroundRED = 0x0c
    __foreGroundYELLOW = 0x0e
    stdOutHandle = ctypes.windll.kernel32.GetStdHandle(__stdOutputHandle)

    def setCmdColor(color, handle=stdOutHandle):
        return ctypes.windll.kernel32.SetConsoleTextAttribute(handle, color)

    def resetCmdColor():
        setCmdColor(__foreGroundRED | __foreGroundGREEN | __foreGroundBLUE)

    def printBlue(msg):
        setCmdColor(__foreGroundBLUE)
        sys.stdout.write(msg + '\n')
        resetCmdColor()

    def printGreen(msg):
        setCmdColor(__foreGroundGREEN)
        sys.stdout.write(msg + '\n')
        resetCmdColor()

    def printRed(msg):
        setCmdColor(__foreGroundRED)
        # sys.stdout.write(msg + '\n')
        print(msg)
        resetCmdColor()

    def printYellow(msg):
        setCmdColor(__foreGroundYELLOW)
        sys.stdout.write(msg + '\n')
        resetCmdColor()

    def pause():
        os.system('pause')

    def locateFile(file):
        print('\n- 即将跳转到输出结果 -\n')
        pause()
        os.system('explorer /select, ' + file)

else:
    STYLE = {
        'fore': {
            'red': 31,
            'green': 32,
            'yellow': 33,
            'blue': 34,
        }
    }

    def UseStyle(msg, mode='', fore='', back='40'):
        fore = '%s' % STYLE['fore'][fore] if fore in STYLE['fore'] else ''
        style = ';'.join([s for s in [mode, fore, back] if s])
        style = '\033[%sm' % style if style else ''
        end = '\033[%sm' % 0 if style else ''
        return '%s%s%s' % (style, msg, end)

    def printRed(msg):
        print(UseStyle(msg, fore='red'))

    def printGreen(msg):
        print(UseStyle(msg, fore='green'))

    def printYellow(msg):
        print(UseStyle(msg, fore='yellow'))

    def printBlue(msg):
        print(UseStyle(msg, fore='blue'))

    def pause():
        pass

    def locateFile(file):
        print('\n- 结果文件保存为：%s' % file)
        print('刷新页面可到 `__output` 文件夹中下载查看。')

if __name__ == '__main__':
    print('****欢迎使用一致性检查工具 %s\n' % __version__)
    print(now(), '**开始运行')
    result = DEBUG_FILE
    try:
        result = initConfig()
    except Exception as err:
        printRed('Error:')
        traceback.print_exc(file=open(DEBUG_FILE, 'w'))
        printRed(err)
        printBlue('出错啦。请反馈目录中的 debug_log.txt 文件内容')
    t = datetime.now() - t0
    print('\n执行用时：%2.4f s' % t.total_seconds())
    # os.system('pause')
    locateFile(result)
